pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1545,6 +1573,304 @@ static pj_status_t create_answer( pj_pool_t *pool,
+            tmp_ret |= 1;
+        }
+        if (pj_strcmp(&sess_med_pt_map->map[i].codec, codec) == 0) {
+            tmp_ret |= 2;
+        }
+
+        if (tmp_ret == 3)
+            return tmp_ret;
+
+        if (tmp_ret > retval)
+            retval = tmp_ret;
+    }
+    return retval;
+}
+
+static media_pt_map* get_media_pt_map(pjmedia_sdp_neg* neg, unsigned idx)
+{
+    unsigned max_cnt = PJ_ARRAY_SIZE(neg->med_pt_map);
+    if (idx >= max_cnt) {
+        return NULL;
+    }
+    if (idx == neg->med_pt_map_cnt)
+        neg->med_pt_map[neg->med_pt_map_cnt++].map_cnt = 0;
+
+    return &neg->med_pt_map[idx];
+}
+
+/*
+ * Update the session pt map. The number of stream will always corresponds to the 
+ * current session. 
+ */
+static void update_session_pt_map(pj_pool_t *pool, const pjmedia_sdp_session* sess,
+                                  pjmedia_sdp_neg* neg)
+{
+    unsigned i = 0;
+    unsigned max_med_cnt = PJ_ARRAY_SIZE(neg->med_pt_map);
+    unsigned med_cnt = 0;
+
+    for (; i < sess->media_count && i < max_med_cnt; ++i) {
+        unsigned j = 0;
+        pjmedia_sdp_media* sess_media = sess->media[i];
+        media_pt_map* med_pt_map = get_media_pt_map(neg, i);
+
+        if (!med_pt_map) {
+            PJ_LOG(4, (THIS_FILE, "Unable to find media pt map for [%.*]", 
+                       sess->media[i]->desc.media.slen, sess->media[i]->desc.media.ptr));
+            break;
+        }
+        ++med_cnt;
+
+        for (; j < sess_media->desc.fmt_count; ++j) {
+            /* Get the codec id */
+            const pjmedia_sdp_attr* attr;
+            pj_str_t codec;
+            unsigned pt = pj_strtoul(&sess_media->desc.fmt[j]);
+            unsigned pt_len;
+
+            if (pt < START_DYNAMIC_PT)
+                continue;
+
+            attr = pjmedia_sdp_media_find_attr2(sess_media, "rtpmap",
+                                                &sess_media->desc.fmt[j]);
+
+            if (!attr) {
+                continue;
+            }
+            pt_len = sess_media->desc.fmt[j].slen + 1;
+            codec.ptr = attr->value.ptr + pt_len;
+            codec.slen = attr->value.slen - pt_len;
+
+            if (pt_exists(med_pt_map, pt, &codec) != 3) {
+                /* If the same PT and codec combination found, don't update the map.
+                 * It is possible remote send:
+                 * - new PT and codec map
+                 * - already assigned PT used by different codec
+                 * - different/new PT used by the same codec
+                 * Later when we send new offer, we can find the PT based on the codec.
+                 */
+                med_pt_map->map[med_pt_map->map_cnt].pt = pt;
+                pj_strdup(pool, &med_pt_map->map[med_pt_map->map_cnt++].codec, &codec);
+            }
+        }
+    }
+    neg->med_pt_map_cnt = med_cnt;
+}
+
+/*
+ * The pt map is maintain using temporary/active pool. This is to avoid memory usage
```

Comment:
This is a good feature, but it seems that all the codec string duplication can be quite inefficient.

Since the number of codec strings should be quite constant (especially if we only record our endpoint), i.e. it will only include all the codecs we support and won't grow beyond that, perhaps we can just store the codec strings using the main pool.